### ISSUE TITLE
[MCHECKSTYLE-59] Checkstyle Report fails if propertyExpansion contains backslash characters

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-checkstyle-plugin#409]
**[Carsten Karkola](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=steinchen)** opened **[MCHECKSTYLE-59](https://issues.apache.org/jira/browse/MCHECKSTYLE-59?redirect=false)** and commented

The chechkstyle plugin fails on windows plattforms, if the propertyExpansion element contains file names like
mms.suppressions.file=${basedir}/../../project/resources/checkstyle/suppressions.xml
The backslashes in ${basedir} (like D:\foo\bar) are missing (D:foobar) and checkstyle fails. It seems to be an issue with StringInputStream.

I've replaced the following old code in CheckstyleReport.java

```
if ( StringUtils.isNotEmpty( propertyExpansion ) )
{
    p.load( new StringInputStream( propertyExpansion ) );
}
```

with the new Code

```
      if ( StringUtils.isNotEmpty( propertyExpansion ) ) {

        BufferedReader reader = new BufferedReader(new StringReader(propertyExpansion));
        String line = reader.readLine();
        while (line != null) {
          int delimPosition = line.indexOf('=');
          if (delimPosition > -1) {
            String name = line.substring(0, delimPosition).trim();
            delimPosition++;
            if (delimPosition < line.length()) {
              String value = line.substring(delimPosition).trim();
              if (value.length() > 0) {
                p.put(name, value);
                getLog().info("Property: " + name + " = " + p.getProperty(name));
              }
            }
          }
          line = reader.readLine();
        }
        reader.close();
      }
```

and it works.

Regards, carsten

---

**Affects:** 2.1

**Attachments:**
- [MCHECKSTYLE-59-PATCH.patch](https://issues.apache.org/jira/secure/attachment/12714699/MCHECKSTYLE-59-PATCH.patch) (_1.29 kB_)
- [MCHECKSTYLE-59-PATCH2.patch](https://issues.apache.org/jira/secure/attachment/12714667/MCHECKSTYLE-59-PATCH2.patch) (_691 bytes_)

6 votes, 7 watchers

**Comment by jira-importer** [comment]
**[Joakim Erdfelt](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=joakime)** commented

Seeing mms.suppressions.file=${basedir}/../../project/resources/checkstyle/suppressions.xml

I can see that you have a multi-module project, and you are attempting to set up a project relative xml definition.
This is undesired, and unsupported by the maven 2.0 checkstyle plugin.

Use the advice provided in http://maven.apache.org/plugins/maven-checkstyle-plugin/customize.html to setup a Custom Checkstyle build tools jar.
This will provide maximum reuse and proper isolation between components.

As well as being supported by continuous integration servers.

Using the build tools advice, and a parent pom, will result in less work for yourself.

**Comment by jira-importer** [comment]
**[Carsten Karkola](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=steinchen)** commented

Thank you for your hint, I tried the suggested solution. In the parent pom I have:

    <suppressionsLocation>de/mms_dresden/checkstyle/suppressions.xml</suppressionsLocation>
    <headerLocation>de/mms_dresden/checkstyle/mms_header_regexp.txt</headerLocation>

On the console I get:
...
[INFO] Generate "Checkstyle" report.
[DEBUG] resolveLocation(de/mms_dresden/checkstyle/t_cruiting_checks.xml, checkstyle-checker.xml)
[DEBUG] Location is not a URL.
[DEBUG] Location is not a File.
[DEBUG] Potential Resource: jar:file:/D:/ckk/m2/de/telekom/mms-dresden/checkstyle/1.0/checkstyle-1.0.jar!/de/mms_dresden
/checkstyle/t_cruiting_checks.xml
[DEBUG] resolveLocation(null, checkstyle-checker.properties)
[DEBUG] resolveLocation(de/mms_dresden/checkstyle/mms_header_regexp.txt, checkstyle-header.txt)
[DEBUG] Location is not a URL.
[DEBUG] Location is not a File.
[DEBUG] Potential Resource: jar:file:/D:/ckk/m2/de/telekom/mms-dresden/checkstyle/1.0/checkstyle-1.0.jar!/de/mms_dresden
/checkstyle/mms_header_regexp.txt
[DEBUG] resolveLocation(null, checkstyle-packages.xml)
[INFO] ------------------------------------------------------------------------
[ERROR] BUILD ERROR
[INFO] ------------------------------------------------------------------------
[INFO] Error during page generation

Embedded error: Error rendering Maven report: Failed during checkstyle configuration
Property ${checkstyle.suppressions.location} has not been set
[INFO] ------------------------------------------------------------------------

The header variable is resolved, but the suppresions variable not?

**Comment by jira-importer** [comment]
**[Lars Behnke](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=behla)** commented

Please find the attached patch. It resolves the issue Carsten pointed out by quoting all backslash characters of the propertyExpansion value. Hence, expressions like \<propertyExpansion>dir=${basedir}\</propertyExpansion> will work on Windows systems. However, there is also a trade-off: You won't be able to pass unicode characters such as \u20ac anymore (which is a minor problem IMO).

**Comment by jira-importer** [comment]
**[Leif Bennett](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=leifbennett)** commented

The problem appears to occur because of the documented behavior of java.util.Properties.load(InputStream), which interprets all backslash characters as escape sequences, and silently drops the backslash if it is not part of a valid escape sequence. On a Windows machine, the propertyExpansion field will contain backslash characters as part of the directory path in 'resourcedir'. These are either removed or changed (e.g., a path containing "plugin\trunk" will convert to "plugin\<tab>runk").

The patch I supply appears to have identical functionality to the one submitted by Lars Behnke, but is perhaps more terse.

A similar strategy could be needed for some of the other values in this method, but have not caused a problem for me.

**Comment by jira-importer** [comment]
**[Dennis Lundberg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dennisl@apache.org)** commented

I applied the patch from Leif Bennett, but it included Java 1.5+ language features, so I used a similar method from StringUtils instead.

A new SNAPSHOT has been deployed. Please verify that it solves this issue.

**Comment by jira-importer** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

can you please open a new issue (or clone this issue) and a provide a sample project to reproduce this issue.
Thanks
